### PR TITLE
Add choice validate input #34

### PR DESCRIPTION
**Pull request by valentinbujdoso**
@magwas 
Could you explain me this input validation?
String choiceId: id of existing choice in the vote with voteid

So the choiceId is valid when the Vote objects Choice list contains the choiceId?

### CODE REVIEW

**Comment by magwas** [comment]
The underlying assumption is that a Choice does have an id, as the name can change.
See https://adadocs.demokracia.rulez.org/PDEngine/edemo/master/index.html#9a443205-d873-4413-8c2d-d6e1e7e5e520

File: src/main/java/org/rulez/demokracia/pdengine/ChoiceManager.java [hunk 1]
```diff
@@ -9,7 +9,11 @@ public ChoiceManager(WebServiceContext wsContext) {
 	}
 
 	public String addChoice(String adminKey, String voteId, String choiceName, String user) {
-		return getVote(voteId).addChoice(choiceName, user);
+		Vote vote = getVote(voteId);
+		vote.checkAdminKey(adminKey);
+		//if(vote.choices.containsKey(choiceName))
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/ChoiceManager.java]: https://adadocs.demokracia.rulez.org/zenta-tools/magwas/master/zenta-tools/codingrules.html#961c75b2-2030-4f7d-83c2-4c4359028322

File: src/main/java/org/rulez/demokracia/pdengine/ChoiceManager.java [hunk 2]
```diff
@@ -9,7 +9,11 @@ public ChoiceManager(WebServiceContext wsContext) {
...
+			return getVote(voteId).addChoice(choiceName, user);
+		//return "Exception";
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/ChoiceManager.java]: https://adadocs.demokracia.rulez.org/zenta-tools/magwas/master/zenta-tools/codingrules.html#961c75b2-2030-4f7d-83c2-4c4359028322

File: src/test/java/org/rulez/demokracia/pdengine/ChoiceAddValidationTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	@tested_behaviour("validates inputs")
+	@Test
+	public void invalid_adminKey_is_rejected() throws ReportedException {
+		String invalidAdminKey = "invalidAdminKey";
+		assertThrows(
+			() -> voteManager.addChoice(invalidAdminKey, adminInfo.voteId, "choice1", "user")
+		).assertMessageIs(String.format("Illegal adminKey: %s", invalidAdminKey));
+	}
+	
+	//@tested_feature("Manage votes")
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/ChoiceAddValidationTest.java]: https://adadocs.demokracia.rulez.org/zenta-tools/magwas/master/zenta-tools/codingrules.html#961c75b2-2030-4f7d-83c2-4c4359028322

**Review by magwas — CHANGES_REQUESTED** [review]
We also need to test all parameters.
For that I have to be sure what the parameters are. Maybe there is a model error here, please be patient while I figure out the Right Way(TM).

**Comment by magwas** [comment]
Please disregard my above comment about choiceId. Cited wrong operation, and not sure whether the add choice operation is modelled right.

**Comment by magwas** [comment]
The right operation:
POST /choice/<voteId>/<adminKey> (String choice) -> String choiceId
So no choiceId in parameters.
choice should still have to be input validated.
Updating the model shortly.

**Comment by valentinbujdoso** [comment]
POST /choice/'voteId'/'adminKey' (String choice) -> String choiceId

so this way, right?

@magwas The choice will be name of the Choice and the adminKey will be the username for the Choice, right?

**Comment by magwas** [comment]
> POST /choice/'voteId'/'adminKey' (String choice) -> String choiceId
> 
> so this way, right?

Yep, sorry for the silly markup.

> 
> @magwas The choice will be name of the Choice and the adminKey will be the username for the Choice, right?

The adminKey is a per-vote key, which can be used to administer the vote. This have been choosen to be made possible to create a vote without logging in. It is different from the username.
The username is planned to be injected by the controller code. This way there is no dependency to WS concepts in the code.

**Comment by valentinbujdoso** [comment]
I have 2 question:
1st: I do not understand how can I create Choice without username. 
2nd: The choice will be a valid input if the votes choices contain the choice, right?

Edit: I found out the first one with getWsUserName(); function.

**Review by magwas — APPROVED** [review]

**Comment by magwas** [comment]
testcaseCount=3

**Pull request merged**